src/main.rs

```diff
@@ -0,0 +1,167 @@
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[allow(non_snake_case)]
+struct UsageId {
+  Id: u16,
+  Name: String,
+  Kinds: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[allow(non_snake_case)]
+struct UsagePage {
+  Kind: String,
+  Id: u16,
+  Name: String,
+  UsageIds: Vec<UsageId>,
+  #[serde(skip)]
+  _UsageIdGenerator: Map<String, Value>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[allow(non_snake_case)]
+struct HidUsageTable {
+  UsageTableVersion: u8,
+  UsageTableRevision: u8,
+  UsageTableSubRevisionInternal: u8,
+  LastGenerated: String,
+  UsagePages: Vec<UsagePage>,
+}
+
+fn usage_tables() -> HidUsageTable {
+  let file = File::open(concat!(env!("CARGO_MANIFEST_DIR"), "/resources/", "HidUsageTables.json"))
+    .expect("Failed to open HidUsageTable JSON");
+  serde_json::from_reader(BufReader::new(file)).expect("Error parsing HidUsageTable JSON")
+}
+
+fn main() {
+  let args = Arguments::parse();
+
+  let raw_descriptor = fs::read(args.path).expect("Failed to read raw descriptor");
+
+  let parsed_descriptor = parse_report_descriptor(&raw_descriptor).expect("Failed to parse descriptor.");
+
+  let mut reports: Vec<_> = std::iter::repeat(ReportType::Input).zip(parsed_descriptor.input_reports).collect();
+  reports.extend(std::iter::repeat(ReportType::Output).zip(parsed_descriptor.output_reports));
+  reports.extend(std::iter::repeat(ReportType::Feature).zip(parsed_descriptor.features));
+
+  let filtered_reports = reports.iter().filter(|(report_type, report)| {
+    if let Some(requested_type) = &args.report_type {
+      if requested_type != report_type {
+        return false;
+      }
+    }
+    if let Some(requested_id) = args.report_id {
+      match report.report_id {
+        Some(id) => {
+          let id: u32 = id.into();
+          if id != requested_id {
+            return false;
+          }
+        }
+        None => return false,
+      }
+    }
+    true
+  });
+
+  let ut = usage_tables();
+
+  for (report_type, report) in filtered_reports {
+    println!("{report_type:?} id: {:?}", report.report_id);
+    for field in &report.fields {
+      match field {
+        ReportField::Variable(v) => {
+          if v.bits.len() == 1 {
+            print!("\tbit:  {:?}", v.bits.start);
+          } else {
+            print!("\tbits: {:?}..{:?}", v.bits.start, v.bits.end - 1);
+          }
+          print!("\tusage: ({:?}:{:?})\t- ", v.usage.page(), v.usage.id());
+          let usage_page = ut.UsagePages.iter().find(|x| x.Id == v.usage.page());
+          if let Some(page) = usage_page {
+            print!(" page: {:?} ", page.Name);
+            let usage = page.UsageIds.iter().find(|x| x.Id == v.usage.id());
+            if let Some(usage) = usage {
+              print!("id: {:?}", usage.Name);
+            } else {
+              print!("id: <unrecognized>");
+            }
+          } else {
+            print!("page: <unrecognized>")
+          }
+          println!("")

```

Comment:
You can drop the empty string.

```suggestion
          println!()
`````

Reply:
Fixed as part of clippy cleanup.